src/Evolution/DiscontinuousGalerkin/DgElementArray.hpp

```diff
@@ -40,12 +43,12 @@ CREATE_HAS_STATIC_MEMBER_VARIABLE_V(use_z_order_distribution)
  * `PhaseDepActionList`.
  *
  * The element assignment to processors is performed by
- * `domain::BlockZCurveProcDistribution` (using a Morton space-filling curve),
- * unless `static constexpr bool use_z_order_distribution = false;` is specified
- * in the `Metavariables`, in which case elements are assigned to processors via
- * round-robin assignment. In both cases, an unordered set of `size_t`s can be
- * passed to the `allocate_array` function which represents physical processors
- * to avoid placing elements on.
+ * `domain::WeightedBlockZCurveProcDistribution` (using a Morton space-filling
```

Comment:
Has this been tested for global time-stepping?  It's not obvious that the weighted version will be better than the unweighted one in that case, and I think GTS the most commonly used evolution mode.  Perhaps we should keep the old behavior be default and make the weighted version opt-in, or at least give an option to opt out.

Reply:
I have not tested this for GTS, but I'm happy to make it so that it only uses the weighted case it we're using LTS, for now while we don't know its impact when using GTB. Would I just need to check the value of `local_time_stepping` in the Metavariables or is there more to it?

Reply:
Pretty much.  If there is an easy way to template this code so the choice can be made independently from the time stepping that might be easier for people to experiment with.

Reply:
Just want to make sure I understand correctly. My current understanding is that Metavariables for some evolution has a `static constexpr use_local_time_stepping`, like in `EvolveGhBinaryBlackHole.hpp`. I could just use an `if constexpr` in `DgElementArray.hpp` to check `Metavariables::use_local_time_stepping` and do the unweighted distribution if it's false and use the weighted if it's true. However, since this value is set as `true` in `EvolveGhBinaryBlackHole.hpp`, that sounds like BBH runs would always use the weighted distribution. Since you're saying that you want people to have the option to opt-in to the weighted distribution, should I be deciding  which distribution to use based on something else that the user specifies in the yaml? Or do I have a misunderstanding of how/when LTS/GTS is decided for an executable?

Reply:
I think Will means instead of using time stepping (and the `metavars::use_local_time_stepping` bool) as the determining factor for whether to do weighted vs unweighted distributions, just add a template bool to `DgElementArray` called `UseWeightedZCurve` and use that (maybe default it to false??). That way if, for some reason, we need to turn off weighted distributions (for any executable, not just BBH) we don't also have to turn off LTS.

Reply:
Ok I can add that tparam to `DgElementArray`. In that case, should I just make the default be `false` and only update `EvolveGhBinaryBlackHole` metavars to use `true`? Or are there other executable metavars that I should also tell to use the weighted distribution?

Reply:
If you you're up for a bunch of mindless file updating, setting it to `use_local_time_stepping` everywhere is probably a good start.  While you've only tested on BBH, I would be surprised if this was not a big improvement on any LTS evolution.

I'm envisioning that we may end up with multiple distribution algorithms and this option could eventually evolve into some enum or something.  Maybe a distribution weighted by the number of points but leaving out the spacing factor, for example.  Or something that can handle nonuniform refinement.

Reply:
Sure, I'll do that! I agree with the vision for a few different distributions, one of them being just based on the number of grid points

Reply:
For elliptic solves I think the weight should be based only on the number of points, not on the spacing factor (because there is no CFL condition), so allowing for different distributions would be very helpful. Passing in a runtime(!) bool to disable the spacing factor would be enough for now and probably quite easy to add. Making the bool an enum with cases `Weight::Uniform`, `Weight::MinGridSpacing`, `Weight::NumPoints` would probably also be quite easy and relatively future-proof, so I suggest that. Does that good to everyone?

Reply:
The elliptic DgElementArray still uses the original element distribution. That wasn't changed in the PR. I recommend we hold off on adding even more things to this PR besides the already long list of change requests (and the template bool below). A follow up PR can do run-time things, add `enum`s, and potentially edit the elliptic DgElementArray if needed.

Reply:
No need to change the elliptic DgElementArray in this PR of course (I didn't request that), but supporting different weight functions is a design question for this PR. Doesn't have to be implemented in this PR, but we should have a plan for the implementation. We don't want to refactor this completely later to support a feature we surely need.

Reply:
Ah ok I understand now. Yeah an option-creatable `enum` seems good as an alternative to the template bool.

Reply:
@nilsvu So I understand what is being asked of for this PR (so I can push fixups), which function are you suggesting take a bool (since I don't think I'm allowed to change the signature of `DgElementArray::allocate_array()`)? Or based on what Kyle said, am I being asked to create a new `Option` that is parsed from the yaml files and grabbed from the cache or something? I understand we are wanting to be forward-thinking about the design to accommodate features we want to add, I'm just unclear what I'm being asked to do for this PR regarding the runtime bool

Reply:
Can you just make the constructor of the element distribution take a bool/enum? That would be enough I think. Option-parsing etc can be done later.

Reply:
I can do that, but then I think that arg for `WeightedElementDistribution` would only make sense for specifying `Weight::NumPointsOverSqrtMinGridSpacing` or `Weight::NumPoints`, while the `Weight::Uniform` one wouldn't make sense as an option for this constructor since we instead construct `ElementDistribution` to do that, meaning if we want to have logic that decides between weighted (`WeightedElementDistribution`) vs unweighted (`ElementDistribution`), I think we need that bool/enum  checked before we decide which distribution to construct between weighted and unweighted. Either that or we need to merge the two distribution classes, maybe. Let me know what you think and if this made sense. Basically, I can do what you're suggesting, I just think that doesn't cover Will's request to make weighted/unweighted toggleable

Reply:
I thought that the weighted distribution just replaces the unweighted, and reduces to the unweighted for the `Weight::Uniform` enum case. If deleting the unweighted distribution is too much work for this PR then don't worry about it, but I thought that was the plan?